File: frontend/src/models/task.ts
// frontend/src/models/task.ts
// Task entity model for the AIDO task management application

export interface Task {
  id: number; // Backend uses integer IDs, not strings
  created_by: number; // Backend uses 'created_by', not 'user_id'
  title: string;
  description?: string;
  priority: 'low' | 'medium' | 'high' | 'urgent'; // Backend has 4 priority levels
  status: 'todo' | 'in_progress' | 'done' | 'blocked'; // Backend uses status enum, not completed boolean
  assigned_to?: number;
  project_id?: number;
  due_date?: string; // ISO 8601 date string
  created_at: string; // ISO 8601 date string
  updated_at: string; // ISO 8601 date string
  // Remove completed field - use status instead
  estimated_time?: number; // in minutes
  tags?: string[]; // array of tag IDs
  subtasks?: Subtask[];
  category?: string;
  position?: number; // for ordering tasks
  reminder_time?: string;
}

export interface Subtask {
  id: string;
  task_id: string;
  title: string;
  completed: boolean;
  order_index: number;
  created_at: string;
  updated_at: string;
}

export interface TaskCreateRequest {
  title: string;
  description?: string;
  priority?: 'low' | 'medium' | 'high' | 'urgent'; // Backend has 4 priority levels
  status?: 'todo' | 'in_progress' | 'done' | 'blocked'; // Backend uses status enum
  assigned_to?: number;
  project_id?: number;
  due_date?: string; // ISO 8601 date string
  estimated_time?: number; // in minutes
  tags?: string[]; // array of tag IDs
  category?: string;
  position?: number; // for ordering tasks
  reminder_time?: string;
}

export interface TaskUpdateRequest {
  title?: string;
  description?: string;
  priority?: 'low' | 'medium' | 'high' | 'urgent'; // Backend has 4 priority levels
  status?: 'todo' | 'in_progress' | 'done' | 'blocked'; // Backend uses status enum
  assigned_to?: number;
  project_id?: number;
  due_date?: string; // ISO 8601 date string
  estimated_time?: number; // in minutes
  tags?: string[]; // array of tag IDs
  category?: string;
  position?: number; // for ordering tasks
  reminder_time?: string;
}

export interface TaskFilterOptions {
  skip?: number;
  limit?: number;
  // Update filters to match backend expectations
  priority?: 'low' | 'medium' | 'high' | 'urgent';
  status?: 'todo' | 'in_progress' | 'done' | 'blocked';
  due_date?: string; // ISO 8601 date string
  tags?: string[]; // array of tag IDs
  category?: string;
  search?: string; // search term for title/description
  start_date?: string; // ISO 8601 date string
  end_date?: string; // ISO 8601 date string
}

export interface TaskListResponse {
  tasks: Task[]; // Update to match actual backend response
  total?: number;
  page?: number;
  limit?: number;
}

File: backend/src/models/schemas/task.py
from pydantic import BaseModel, field_validator
from typing import Optional
from datetime import datetime
from enum import Enum


class PriorityEnum(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class StatusEnum(str, Enum):
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    DONE = "done"
    BLOCKED = "blocked"


class TaskBase(BaseModel):
    title: str
    description: Optional[str] = None
    priority: PriorityEnum = PriorityEnum.MEDIUM
    status: StatusEnum = StatusEnum.TODO
    assigned_to: Optional[int] = None
    project_id: Optional[int] = None
    due_date: Optional[datetime] = None

    @field_validator('title')
    @classmethod
    def validate_title(cls, v):
        if not v or len(v) < 1 or len(v) > 200:
            raise ValueError('Title must be between 1 and 200 characters')
        return v


class TaskCreate(TaskBase):
    # According to the requirements, only title and priority are required for task creation
    title: str
    priority: PriorityEnum = PriorityEnum.MEDIUM


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[PriorityEnum] = None
    status: Optional[StatusEnum] = None
    assigned_to: Optional[int] = None
    project_id: Optional[int] = None
    due_date: Optional[datetime] = None

    @field_validator('title')
    @classmethod
    def validate_title(cls, v):
        if v is not None and (len(v) < 1 or len(v) > 200):
            raise ValueError('Title must be between 1 and 200 characters')
        return v


class TaskResponse(TaskBase):
    id: int
    created_by: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True